### PR TITLE
basic_direct play

### PR DESCRIPTION
**Pull request by MomenYacoub**

### CODE REVIEW

File: soccer/gameplay/plays/restarts/basic_direct.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    def on_enter_kicking(self):
+        self.add_subbehavior(skills.pivot_kick.PivotKick(),'kicking_the_ball')
+
+    def on_enter_passing(self):
+        kicker = skills.pivot_kick.PivotKick()
+        kicker.use_chipper = True
+        kicker.kick_power = 50
+        receive_pt = robocup.Point(0, 3 * constants.Field.Length / 4)
+
+        chip_behavior = tactics.coordinated_pass.CoordinatedPass(receive_pt,None,(kicker, lambda x : True), None, False, False)
```
Comment (Jason27chan) [inline soccer/gameplay/plays/restarts/basic_direct.py]: Format code more like https://github.com/RoboJackets/robocup-software/blob/master/soccer/gameplay/plays/restarts/our_free_kick.py#69 to make it easier to read

File: soccer/gameplay/plays/restarts/basic_direct.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+import tactics.coordinated_pass
+import skills
+
+
+class DirectFreeKick(standard_play.StandardPlay):
+
+    class State(enum.Enum):
+        kicking = 1
+        passing = 2
+    def __init__ (self):
```
Comment (Jason27chan) [inline soccer/gameplay/plays/restarts/basic_direct.py]: Add a newline between the state class and __init__ function please. This would be fixed by make pretty :(

File: soccer/gameplay/plays/restarts/basic_direct.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        super().__init__(continuous = True)
+
+        self.add_state(DirectFreeKick.State.kicking,behavior.Behavior.State.running)
+        self.add_state(DirectFreeKick.State.passing,behavior.Behavior.State.running)
+
+
```
Comment (Jason27chan) [inline soccer/gameplay/plays/restarts/basic_direct.py]: delete whitespace. Would be fixed by make pretty

File: soccer/gameplay/plays/restarts/basic_indirect.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
+import standard_play
```
Comment (Jason27chan) [inline soccer/gameplay/plays/restarts/basic_indirect.py]: This file shouldn't appear in the diff after we merge my changes and you pull again.

File: soccer/gameplay/plays/restarts/basic_direct.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+class DirectFreeKick(standard_play.StandardPlay):
+
+    class State(enum.Enum):
+        kicking = 1
+        passing = 2
+    def __init__ (self):
+        super().__init__(continuous = True)
+
+        self.add_state(DirectFreeKick.State.kicking,behavior.Behavior.State.running)
+        self.add_state(DirectFreeKick.State.passing,behavior.Behavior.State.running)
```
Comment (JNeiger) [inline soccer/gameplay/plays/restarts/basic_direct.py]: This can be condensed into a for loop.

```
        for s in DirectFreeKick.State:
            self.add_state(s, behavior.Behavior.State.running)
```

File: soccer/gameplay/plays/restarts/basic_direct.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+            lambda:evaluation.shooting.eval_shot(main.ball().pos, main.our_robots()) <0.35,'Pass')
+
+    def on_enter_kicking(self):
+        self.add_subbehavior(skills.pivot_kick.PivotKick(),'kicking_the_ball')
+
+    def on_enter_passing(self):
+        kicker = skills.pivot_kick.PivotKick()
+        kicker.use_chipper = True
+        kicker.kick_power = 50
+        receive_pt = robocup.Point(0, 3 * constants.Field.Length / 4)
```
Comment (JNeiger) [inline soccer/gameplay/plays/restarts/basic_direct.py]: The 3/4 should be a varible at the top

File: soccer/gameplay/plays/restarts/basic_corner_kick.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+		#Depending on corner's location the point are refelcted
+		if main.ball().pos.x < 0:
+			refelctor = -1
+
+		line1 = PointAndPriority(robocup.Line(robocup.Point(-1.9 * refelctor, 7.45), 
+			robocup.Point(-1.9 * refelctor, 8.5)), 0)
+		line2 = PointAndPriority(robocup.Line(robocup.Point(-0.56 * refelctor, 7.11), 
+			robocup.Point(0.44 * refelctor, 6.2)), 1)
+		line3 = PointAndPriority(robocup.Line(robocup.Point(2.40 * refelctor, 5.85), 
+			robocup.Point(0.99 * refelctor, 5.85)), 2)
```
Comment (JNeiger) [inline soccer/gameplay/plays/restarts/basic_corner_kick.py]: These should be % of field width and length. Our field in the shop is a different size than the default B field which is a different size than the default A field. This allows for the same code to be used in all the cases without messing stuff up

File: soccer/gameplay/plays/restarts/basic_corner_kick.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+		lines = [line1, line2, line3]
+
+		#Takes three points on the line (the two ends and the midpoint) and returns one with highest probability
+		def best_point_on_line(PointAndPriority):
+			line = PointAndPriority.line_pt
+			point1 = line.get_pt(0)
+			point2 = line.get_pt(1)
+			point3_x = (point1.x + point2.x) * 0.5
+			point3_y = (point1.y + point1.y) * 0.5
+			point3 = robocup.Point(point3_x, point3_y)
```
Comment (JNeiger) [inline soccer/gameplay/plays/restarts/basic_corner_kick.py]: I think this can be simplified to just `0.5 * (point1 + point2)

File: soccer/gameplay/plays/restarts/basic_corner_kick.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+			robocup.Point(-1.9 * refelctor, 8.5)), 0)
+		line2 = PointAndPriority(robocup.Line(robocup.Point(-0.56 * refelctor, 7.11), 
+			robocup.Point(0.44 * refelctor, 6.2)), 1)
+		line3 = PointAndPriority(robocup.Line(robocup.Point(2.40 * refelctor, 5.85), 
+			robocup.Point(0.99 * refelctor, 5.85)), 2)
+
+		lines = [line1, line2, line3]
+
+		#Takes three points on the line (the two ends and the midpoint) and returns one with highest probability
+		def best_point_on_line(PointAndPriority):
```
Comment (JNeiger) [inline soccer/gameplay/plays/restarts/basic_corner_kick.py]: @Jason27chan may have some more thoughts on this, but personally, I like to avoid functions defined inside other functions (if they are more than a line or two). It makes it a little hard to follow the code in some cases. I still have flashbacks to refactoring the defense to remove all of that.

This is the [old defense](https://github.com/RoboJackets/robocup-software/blob/master/soccer/gameplay/tactics/defense_old.py), this is the [new one](https://github.com/RoboJackets/robocup-software/blob/master/soccer/gameplay/tactics/defense.py)

File: soccer/gameplay/plays/restarts/basic_corner_kick.py [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+	def __init__(self, line_pt, priority):
+		self.line_pt = line_pt
+		self.priority = priority
+
+class BasicCorner(standard_play.StandardPlay):
+	def __init__(self):
+		super().__init__(continuous = True)
+		#the minimum number we want the probability multiplier be
+		MINIMUM_MULTIPLIER = 0.5
+		refelctor = 1
```
Comment (JNeiger) [inline soccer/gameplay/plays/restarts/basic_corner_kick.py]: refelctor -> reflector

File: soccer/gameplay/plays/restarts/basic_corner_kick.py [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+class PointAndPriority:
+	def __init__(self, line_pt, priority):
+		self.line_pt = line_pt
+		self.priority = priority
+
+class BasicCorner(standard_play.StandardPlay):
+	def __init__(self):
+		super().__init__(continuous = True)
+		#the minimum number we want the probability multiplier be
+		MINIMUM_MULTIPLIER = 0.5
```
Comment (JNeiger) [inline soccer/gameplay/plays/restarts/basic_corner_kick.py]: Could you go into a little more depth on what this config value represents? It sounds as if it's the maximum penalty on pass chance between the highest priority and lowest priority segment, but I'm not sure

File: soccer/gameplay/plays/restarts/basic_corner_kick.py (hunk 7 above)
Comment (JNeiger) [inline soccer/gameplay/plays/restarts/basic_corner_kick.py]: In general, are the priorities for these lines going from 0 up to the number of lines - 1?

File: soccer/gameplay/plays/restarts/basic_direct.py [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+    def __init__ (self):
+        super().__init__(continuous = True)
+        for s in DirectFreeKick.State:
+            self.add_state(s, behavior.Behavior.State.running)
+        # If chance of scoring is greater than equal to MIN_SHOT_PROB it will kick
+        self.add_transition(behavior.Behavior.State.start,DirectFreeKick.State.kicking,
+            lambda:evaluation.shooting.eval_shot(main.ball().pos, main.our_robots()) >= MIN_SHOT_PROB,'ShotChanceHigh')        
+        #If chance of scoring is below MIN_SHOT_PROB we want to pass using the chipper
+        self.add_transition(behavior.Behavior.State.start,DirectFreeKick.State.passing,
+            lambda:evaluation.shooting.eval_shot(main.ball().pos, main.our_robots()) < MIN_SHOT_PROB,'Pass')
```
Comment (JNeiger) [inline soccer/gameplay/plays/restarts/basic_direct.py]: I would space these out a little bit, I like the comments and everything though

```
for s in asdf:
    self.add_state(asdf)

# comment
self.add_transition(behavior.Behavior.State.start,
                    DirectFreeKick.State.Kicking,
                    lambda: evaluation.shooting.eval_shot(asdf) >= asdf,
                    'ShotChanceHigh')

# comment
self.add_transition(behavior.Behavior.State.start,
                    DirectFreeKick.State.passing,
                    lambda: evaluation.shooting.eval_shot(asdf) < asdf,
                    'Pass')

Comment (JNeiger) [inline soccer/gameplay/plays/restarts/basic_direct.py]: @Jason27chan would need to confirm, but I'm 90% sure there needs to be a way to go to the `behavior.Behavior.State.completed` or `behavior.Behavior.State.failed` so we can start our normal offence again

File: soccer/gameplay/plays/restarts/basic_direct.py [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    def on_enter_passing(self):
+        kicker = skills.pivot_kick.PivotKick()
+        kicker.use_chipper = True
+        kicker.kick_power = 50
+        #The point the ball will be chipped to
+        receive_pt = robocup.Point(0, FIELD_LENGTH_MULTIPLIER * constants.Field.Length)
+
+        chip_behavior = tactics.coordinated_pass.CoordinatedPass(receive_pt,None,
+            (kicker, lambda x : True), 
+            None, False, False)
```
Comment (JNeiger) [inline soccer/gameplay/plays/restarts/basic_direct.py]: I would format this close to what is done here. It makes it a little easier to read.
https://github.com/RoboJackets/robocup-software/blob/master/soccer/gameplay/plays/restarts/our_free_kick.py#69

**Comment by JNeiger** [comment]
Looking good! Just a few style things and maybe that `completed/failed` thing. Once those are fix it should be good to merge!

File: soccer/gameplay/plays/restarts/basic_direct.py (hunk 12 above)
Comment (Jason27chan) [reply soccer/gameplay/plays/restarts/basic_direct.py]: Yes we do need those transitions!

File: soccer/gameplay/plays/restarts/basic_direct.py [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+    MIN_SHOT_PROB = 0.35
+    FIELD_LENGTH_MULTIPLIER = 0.75
+
+    class State(enum.Enum):
+        kicking = 1
+        passing = 2
+
+    def __init__ (self):
+        super().__init__(continuous = True)
+        for s in DirectFreeKick.State:
```
Comment (Jason27chan) [inline soccer/gameplay/plays/restarts/basic_direct.py]: Class is no longer called DirectFreeKick, so need to change names here and in behaviors

File: soccer/gameplay/plays/restarts/basic_direct.py [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+            self.add_state(s, behavior.Behavior.State.running)
+        # If chance of scoring is greater than equal to MIN_SHOT_PROB it will kick
+        self.add_transition(behavior.Behavior.State.start,DirectFreeKick.State.kicking,
+            lambda:evaluation.shooting.eval_shot(main.ball().pos, main.our_robots()) >= MIN_SHOT_PROB,'ShotChanceHigh')
```
Comment (Jason27chan) [inline soccer/gameplay/plays/restarts/basic_direct.py]: MIN_SHOT_PROB undefined. Either use BasicDirect.MIN_SHOT_PROB or make it an instance variable, same for FIELD_LENGTH_MULTIPLIER

**Comment by tjones320** [comment]
Does this branch need to stay open? I feel like nearly all commits are the same as the other branch? @JNeiger 
@Jason27chan

**Comment by JNeiger** [comment]
The other PR has this file plus a few new ones. I was going to delete this one file out of the other PR to keep each one "under" a single topic since they aren't directly related

**Comment by kylestach** [comment]
@MomenYacoub, what is the status here? Is this still alive?
